### DISCUSSION TITLE
BeforeTestRun BoDi Dependency Injection bug?

### DISCUSSION THREAD

**Discussion by bethly-hallows** [reqnroll/Reqnroll#88] [Q&A] [answered]
Hi All,

I have recently picked up investigating migrating our test projects from Specflow to Reqnroll but have happened upon an issue, and unsure if this is a known issue/bug or if I am simply missing something. As this is a new project I figured this would be the best place to ask!

When using Specflow, I am able to register dependencies in the [BeforeTestRun] hook, as below. Here, I am just registering an instance of my test settings, as an example:

![image](https://github.com/reqnroll/Reqnroll/assets/97543198/142b5761-cbbd-48a2-9630-1ec5499b867b)

Then, say I want to use the test settings in a [BeforeScenario] hook, I am able to inject them via a constructor:

![image](https://github.com/reqnroll/Reqnroll/assets/97543198/ab82b30d-86e3-4de2-9b15-f157103a980f)

This method I have used time and time again, with no issue with the constructor happily injecting that dependency for me from within that container instance:

![image](https://github.com/reqnroll/Reqnroll/assets/97543198/650afa31-0aad-47c4-ae46-7c6ca6c557e0)

Here is the issue.... when I migrate to Reqnroll (with absolutely no other changes except the package and Using reference), the container instance does not seem to carry over to the [BeforeScenario] hooks!

![image](https://github.com/reqnroll/Reqnroll/assets/97543198/8fb89e78-6db1-4d17-a7a6-0d8ca24071c3)

I discovered the issue in an existing test project with many dependencies, but have replicated it on a brand new Reqnroll project with the test settings for ease.

Here are my project files with Specflow and with Reqnroll, for reference:

![image](https://github.com/reqnroll/Reqnroll/assets/97543198/d088d989-4264-4c40-983e-851ebe68ca22)
![image](https://github.com/reqnroll/Reqnroll/assets/97543198/de332a32-c5dc-490c-a0c1-57bfb9b760a8)

Would appreciate any thoughts on this, in the existing test project we have been doing it this way for years and the migration documentation didn't mention anything differently regarding DI.

**Comment by ajeckmans** [comment]
If you want to remain test framework independent you should not create a dependency between BeforeTestrun and BeforeScenario as they're not guaranteed to run in the same thread. Just a note of warning as I don't think this approach will work with all test runners.

Other than that I remember something in the documentation hinting at this not being a supported approach, but I can't find it anymore. Either I'm mistaken or the warning was removed at some point in the last two years.

This is touching on some work that I'm doing to make it possible to resolve scenario specific value retrievers, which also means properly keeping track of the container across the various hooks. That change is a bit slow going as I do have some time I can spend at work on it, but it has been given low priority unfortunately. I'll keep in mind to test this scenario as well if it is not resolved before then :)

**Reply by clrudolphi** [reply]
Would you mind posting the example project which recreates this problem?

**Reply by bethly-hallows** [reply]
@ajeckmans so are you suggesting I should be registering specific dependencies where required in the BeforeScenario hooks instead? I would have to duplicate the code to get the appsettings/usersecrets strings I require is the only thing. 

I'll be honest I was suspicious of this as all the examples I could find only seemed to register in the BeforeScenario hook itself, was this changed in a beta version of Specflow and then Reqnroll has branched off that, and that is why I am only now seeing the issue?


@clrudolphi here is the project, currently with Reqnroll installed. 
[BethPlusOne.zip](https://github.com/reqnroll/Reqnroll/files/14849288/BethPlusOne.zip)

**Reply by bethly-hallows** [reply]
Update: Updating to the latest beta of Specflow (4.0.31-beta) produces the same issue. Just unsure now if it is a bug or purposeful.

**Reply by ajeckmans** [reply]
@bethly-hallows No, like I said I remember reading that this was not a valid scenario, but I cannot find anything in the documentation about that anymore, so I must be remembering it wrong.
If I look at the code where the containers are built this should still just work. I'm a bit at a loss as to why the scenario container does not branch off of the feature container, but at least it branches off of the test thread container which in turn branches off of the global container which is the one you get injected in the `BeforeTestRun` hook.

Interestingly I do see I note in the changelog: https://github.com/reqnroll/Reqnroll/blob/main/CHANGELOG.md?plain=1 for vNext about the resolution context surrounding this (which is from this commit. https://github.com/reqnroll/Reqnroll/commit/3c639558447d2a23e22bda5656271d74a79727b0) 
I wonder whether the BeforeTestRun hook is somehow running in a different test thread than your BeforeScenario hook?

**Reply by ajeckmans** [reply]
Right...

> This normally should not matter for single-threaded tests, but it seems that the way how SpecFlow v4 (and Reqnroll) invokes these hooks is different and they are not necessarily invoked from the same test thread that is used for test execution (this depends on the test execution framework probably, **but with NUnit there seem to be a different thread**), so anything that you register in [BeforeTestRun] to the injected container might not be visible for the step definitions or scenario hooks.
(see https://github.com/reqnroll/Reqnroll/pull/59)

So you're definitely running into that problem :)

**Reply by ajeckmans** [reply]
Good thing a fix is incoming. @gasparnagy When is the new version released with this fix?

**Reply by bethly-hallows** [reply]
@ajeckmans brilliant, I don't know why I couldn't find that and the related issue when doing my searches! Thankyou so much for your help :)

**Reply by gasparnagy** [reply]
@ajeckmans The challenge is that we have made a breaking change by touching the BoDi dependency (#80), so if we would release right now, that would need to be v2.0.0. Or we cherry pick all other fixes and release it as v1.1.0. Since this has a workaround, I would vote for fixing #56 , #71 and #81 first and then release a v2.

**Reply by bethly-hallows** [reply]
Hi @gasparnagy, I saw on another thread that a new release is impending- was just wondering if it got decided if the above was going to be in it, or if not, if there's any idea of a time frame for the above?

**Reply by gasparnagy** [reply, marked as answer]
@bethly-hallows yes, this is in. I'm currently preparing the release. And plan to have it (at least a preview) by tomorrow.

**Comment by gasparnagy** [comment]
It is out: https://github.com/orgs/reqnroll/discussions/142

closing this topic
